recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch

```diff
@@ -0,0 +1,534 @@
++	    dm_gpio_set_dir_flags(&gpio, GPIOD_IS_IN) < 0) {
++		pr_err("Cannot get pin %s for M.2 configuration\n", gpio_name);
++		return 0;
++	}
++
++	return dm_gpio_get_value(&gpio);
++}
++
++static void set_pinvalue(const char *gpio_name, const char *label, int value)
++{
++	struct gpio_desc gpio;
++
++	if (dm_gpio_lookup_name(gpio_name, &gpio) < 0 ||
++	    dm_gpio_request(&gpio, label) < 0 ||
++	    dm_gpio_set_dir_flags(&gpio, GPIOD_IS_OUT) < 0) {
++		pr_err("Cannot set pin %s for M.2 configuration\n", gpio_name);
++		return;
++	}
++	dm_gpio_set_value(&gpio, value);
++}
++
++static bool board_is_m2(void)
++{
++	struct iot2050_info *info = IOT2050_INFO_DATA;
++
++	return IS_ENABLED(CONFIG_TARGET_IOT2050_A53_PG2) &&
++		info->magic == IOT2050_INFO_MAGIC &&
++		strcmp((char *)info->name, "IOT2050-ADVANCED-M2") == 0;
++}
++
+ static bool board_is_advanced(void)
+ {
+ 	struct iot2050_info *info = IOT2050_INFO_DATA;
+@@ -103,6 +213,8 @@ void set_board_info_env(void)
+ 	if (board_is_advanced()) {
+ 		if (IS_ENABLED(CONFIG_TARGET_IOT2050_A53_PG1))
+ 			fdtfile = "ti/k3-am6548-iot2050-advanced.dtb";
++		else if(board_is_m2())
++			fdtfile = "ti/k3-am6548-iot2050-advanced-m2.dtb";
+ 		else
+ 			fdtfile = "ti/k3-am6548-iot2050-advanced-pg2.dtb";
+ 	} else {
+@@ -118,6 +230,103 @@ void set_board_info_env(void)
+ 	env_save();
+ }
+ 
++static void m2_overlay_prepare(void)
++{
++#if defined(CONFIG_OF_LIBFDT) && defined(CONFIG_OF_BOARD_SETUP)
++	const char *overlay_path;
++	void *overlay;
++	u64 loadaddr;
++	ofnode node;
++	int ret;
++
++	if (connector_mode == BKEY_PCIEX2_EKEY_NONE)
++		return;
++
++	if (connector_mode == BKEY_PCIE_EKEY_PCIE)
++		overlay_path = "/fit-images/bkey-ekey-pcie-overlay";
++	else
++		overlay_path = "/fit-images/bkey-usb3-overlay";
++
++	node = ofnode_path(overlay_path);
++	if (!ofnode_valid(node))
++		goto fit_error;
++
++	ret = ofnode_read_u64(node, "load", &loadaddr);
++	if (ret)
++		goto fit_error;
++
++	ret = ofnode_read_u32(node, "size", &connector_overlay_size);
++	if (ret)
++		goto fit_error;
++
++	overlay = map_sysmem(loadaddr, connector_overlay_size);
++
++	connector_overlay = malloc(connector_overlay_size);
++	if (!connector_overlay)
++		goto fit_error;
++
++	memcpy(connector_overlay, overlay, connector_overlay_size);
++	return;
++
++fit_error:
++	pr_err("M.2 device tree overlay %s not available,\n", overlay_path);
++#endif
++}
++
++static void m2_connector_setup(void)
++{
++	ulong m2_manual_config = env_get_ulong("m2_manual_config", 10,
++					       CONNECTOR_MODE_INVALID);
++	const char *mode_info = "";
++	struct m2_config_pins config_pins;
++	unsigned int n;
++
++	/* enable M.2 connector power */
++	set_pinvalue("gpio@601000_17", "P3V3_M2_EN", 1);
++	udelay(4 * 100);
++
++	m2_manual_config = env_get_ulong("m2_manual_config", 10,
++					 CONNECTOR_MODE_INVALID);
++	if (m2_manual_config < CONNECTOR_MODE_INVALID) {
++		mode_info = " [manual mode]";
++		connector_mode = m2_manual_config;
```

Comment:
should check range

Reply:
Connector mode can be any value if the user does not read the documentation.

Reply:
Oh, misread as "why are we checking the range" - we are checking it to be >= 0 and < CONNECTOR_MODE_INVALID.